backends/p4tools/modules/testgen/test/z3-solver/constraints.cpp

```diff
@@ -0,0 +1,191 @@
+#include <gtest/gtest.h>
+
+#include <optional>
+#include <vector>
+
+#include "backends/p4tools/common/core/z3_solver.h"
+#include "backends/p4tools/common/lib/variables.h"
+#include "ir/ir-generated.h"
+#include "ir/ir.h"
+#include "ir/irutils.h"
+#include "lib/cstring.h"
+
+namespace Test {
+using ConstraintVector = const std::vector<const Constraint *>;
+
+class Z3SolverSatisfiabilityChecks : public ::testing::Test {
+    P4Tools::Z3Solver solver;
+
+ public:
+    /// Checks whether the result of the solver calculating @param expression matches @param
+    /// expectedResult.
+    void testCheckSat(const ConstraintVector &expression, std::optional<bool> expectedResult) {
+        auto result = solver.checkSat(expression);
+        EXPECT_EQ(result, expectedResult);
+    }
+};
+
+TEST_F(Z3SolverSatisfiabilityChecks, BitVectors) {
+    const auto *eightBitType = IR::getBitType(8);
+    const auto *fooVar = P4Tools::ToolsVariables::getSymbolicVariable(eightBitType, "foo");
+    const auto *barVar = P4Tools::ToolsVariables::getSymbolicVariable(eightBitType, "bar");
+    {
+        auto *expression =
+            new IR::Equ(IR::getConstant(eightBitType, 1), IR::getConstant(eightBitType, 1));
+        ConstraintVector inputExpression = {expression};
+        testCheckSat(inputExpression, true);
+    }
+    {
+        auto *expression =
+            new IR::Equ(IR::getConstant(eightBitType, 1), IR::getConstant(eightBitType, 2));
+        ConstraintVector inputExpression = {expression};
+        testCheckSat(inputExpression, false);
+    }
+    {
+        auto *expression = new IR::Equ(fooVar, IR::getConstant(eightBitType, 1));
+        ConstraintVector inputExpression = {expression};
+        testCheckSat(inputExpression, true);
+    }
+    {
+        auto *expression = new IR::Equ(fooVar, IR::getConstant(eightBitType, 1));
+        auto *constraint1 = new IR::Equ(fooVar, IR::getConstant(eightBitType, 1));
```

Comment:
Is this deliberately the same as `expression`?

Reply:
Yes, the other one was not. Fixed.